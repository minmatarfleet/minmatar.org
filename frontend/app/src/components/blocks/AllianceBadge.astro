---
import { i18n } from '@helpers/i18n'
const { t } = i18n(Astro.url)

import type { AllianceBadgeProps } from '@dtypes/layout_components'
import { get_alliance_by_id } from '@helpers/api.eveonline/alliances'
import type { AllianceEvE } from '@dtypes/api.eveonline'
import { prod_error_messages } from '@helpers/env'

interface Props {
    alliance: AllianceBadgeProps;
}

const {
    alliance,
} = Astro.props

let alliance_eve:AllianceEvE | null = null
let get_eve_alliance_error
if (!alliance?.name) {
    try {
        alliance_eve = await get_alliance_by_id(alliance.id)
    } catch (error) {
        get_eve_alliance_error = (prod_error_messages() ? t('get_eve_alliance_error') : error.message)
    }
}

const alliance_name = alliance?.name ?? (alliance_eve?.name ?? get_eve_alliance_error)

import { get_alliance_logo } from '@helpers/eve_image_server';
import FixedFluid from '@components/compositions/FixedFluid.astro'
---
<FixedFluid
    width='32px'
    gap='var(--space-2xs)'
    class="[ items-center ]"
>
    <picture>
        <img loading="lazy" src={get_alliance_logo(alliance.id)} width="32" height="32" />
    </picture>
    <small>
        <span class="[ text-[var(--highlight)] ]">{alliance_name}</span>
    </small>
</FixedFluid>