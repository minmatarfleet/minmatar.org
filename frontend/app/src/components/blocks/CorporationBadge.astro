---
import { i18n } from '@helpers/i18n'
const { t } = i18n(Astro.url)

import type { CorporationBadgeProps } from '@dtypes/layout_components'
import { get_corporation_by_id } from '@helpers/api.eveonline/corporations'
import type { CorporationEvE } from '@dtypes/api.eveonline'
import { prod_error_messages } from '@helpers/env'

interface Props {
    corporation:    CorporationBadgeProps;
    border?:        boolean;
}

const {
    corporation,
    border = false,
} = Astro.props

let corporation_eve:CorporationEvE | null = null
let get_eve_corporation_error
if (!corporation?.name) {
    try {
        corporation_eve = await get_corporation_by_id(corporation.id)
    } catch (error) {
        get_eve_corporation_error = (prod_error_messages() ? t('get_eve_corporation_error') : error.message)
    }
}

const corporation_name = corporation?.name ?? (corporation_eve?.name ?? get_eve_corporation_error)
const size = corporation?.size ?? 'sm'
const logo_size = (size == 'sm' ? 32 : 64)
const description:string | false = corporation?.description ?? false

import { get_corporation_logo } from '@helpers/eve_image_server';
import FixedFluid from '@components/compositions/FixedFluid.astro'
import Flexblock from '@components/compositions/Flexblock.astro';

const gap = size === 'sm' ? 'var(--space-2xs)' : 'var(--space-s)'
---
<FixedFluid
    width={`${logo_size}px`}
    gap={gap}
    class="[ items-center ]"
>
    <picture class:list={{ border: border }}>
        <img loading="lazy" src={get_corporation_logo(corporation.id, logo_size)} width={logo_size} height={logo_size} />
    </picture>
    <Flexblock gap='0' class="[ w-full ]">
        {size == 'lg' ?
            <span>
                <span class="[ text-[var(--highlight)] max-w-fit ]">{corporation_name}</span>
            </span>
            <slot>
                <small>{description}</small>
            </slot>
            :
            <small class="[ max-w-fit ]">
                <a class="[ text-[var(--highlight)] ]">{corporation_name}</a>
            </small>
        }
    </Flexblock>
</FixedFluid>