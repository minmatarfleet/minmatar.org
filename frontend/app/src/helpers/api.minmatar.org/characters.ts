import type {
    Character,
    CharacterSkillset,
    CharacterAsset,
    CharacterSummary,
    CharacterESITokens,
    CharacterTag
} from '@dtypes/api.minmatar.org'
import { get_error_message } from '@helpers/string'

const API_ENDPOINT =  `${import.meta.env.API_URL}/api/eveonline/characters`

export async function get_characters(access_token:string, primary_character_id:number | null = null) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}${primary_character_id ? `?primary_character_id=${primary_character_id}` : ''}`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ))
        }

        return await response.json() as Character[];
    } catch (error) {
        throw new Error(`Error fetching characters: ${error.message}`);
    }
}

export async function get_character_by_id(access_token:string, character_id:number) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/${character_id}`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ))
        }

        return await response.json() as Character;
    } catch (error) {
        throw new Error(`Error fetching character: ${error.message}`);
    }
}

export async function delete_characters(access_token:string, character_id:number) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/${character_id}`

    console.log(`Requesting DELETE: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            method: 'DELETE',
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `DELETE ${ENDPOINT}`
            ))
        }

        return (response.status === 200);
    } catch (error) {
        throw new Error(`Error deleting characters: ${error.message}`);
    }
}

export async function get_primary_characters(access_token:string) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/primary`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (response.status != 200 && response.status != 404) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ))
        }

        return await response.json() as Character;
    } catch (error) {
        throw new Error(`Error fetching main pilot: ${error.message}`);
    }
}

export async function get_character_skillsets(access_token:string, character_id:number) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/${character_id}/skillsets`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ))
        }

        return await response.json() as CharacterSkillset[];
    } catch (error) {
        throw new Error(`Error fetching character skillsets: ${error.message}`);
    }
}

export async function get_character_assets(access_token:string, character_id:number) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/${character_id}/assets`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ))
        }

        return await response.json() as CharacterAsset[];
    } catch (error) {
        throw new Error(`Error fetching character assets: ${error.message}`);
    }
}

export async function set_primary_characters(access_token:string, character_id:number) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/primary?character_id=${character_id}`

    console.log(`Requesting PUT: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            method: 'PUT',
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `PUT ${ENDPOINT}`
            ))
        }

        return (response.status === 200);
    } catch (error) {
        throw new Error(`Error setting main character: ${error.message}`);
    }
}

export async function get_characters_summary(access_token:string) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/summary`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ))
        }

        return await response.json() as CharacterSummary;
    } catch (error) {
        throw new Error(`Error fetching character summary: ${error.message}`);
    }
}

export async function get_character_esi_token(access_token:string, character_id:number) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/${character_id}/tokens`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ))
        }

        return await response.json() as CharacterESITokens[];
    } catch (error) {
        throw new Error(`Error fetching character summary: ${error.message}`);
    }
}

export async function get_tags(access_token:string) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/tags`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ))
        }

        return await response.json() as CharacterTag[];
    } catch (error) {
        throw new Error(`Error fetching character summary: ${error.message}`);
    }
}

export async function set_character_tags(access_token:string, character_id:number, tags:number[]) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/${character_id}/tags`

    console.log(`Requesting PUT: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            method: 'PUT',
            headers: headers,
            body: JSON.stringify(tags),
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `PUT ${ENDPOINT}`
            ))
        }

        return (response.status === 200);
    } catch (error) {
        throw new Error(`Error setting character tags: ${error.message}`);
    }
}

export async function get_character_tags(access_token:string, character_id:number) {
    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${access_token}`
    }

    const ENDPOINT = `${API_ENDPOINT}/${character_id}/tags`

    console.log(`Requesting: ${ENDPOINT}`)

    try {
        const response = await fetch(ENDPOINT, {
            headers: headers
        })

        // console.log(response)

        if (!response.ok) {
            throw new Error(get_error_message(
                response.status,
                `GET ${ENDPOINT}`
            ))
        }

        return await response.json() as CharacterTag[];
    } catch (error) {
        throw new Error(`Error fetching character summary: ${error.message}`);
    }
}