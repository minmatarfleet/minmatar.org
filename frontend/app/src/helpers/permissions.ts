import { get_user_by_name } from '@helpers/api.minmatar.org/authentication'

export async function get_user_permissions(user_name:string) {
    try {
        const profile = await get_user_by_name(user_name)
        return profile?.permissions ?? []
    } catch (error) {
        return []
    }
}