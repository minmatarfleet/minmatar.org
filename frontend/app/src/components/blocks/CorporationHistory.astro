---
import { i18n } from '@helpers/i18n'
const { lang, t, translatePath } = i18n(Astro.url)

import { marked } from 'marked';

import { get_corporation_history } from '@helpers/api.eveonline/characters'
import type { CorporationHistoryItem } from '@dtypes/layout_components'
import type { CorporationHistoryEvE } from '@dtypes/api.eveonline'
import { prod_error_messages } from '@helpers/env'
import { from_to_text, from_to_now_text } from '@helpers/date'
import type { NamesAndCategoriesEvE } from '@dtypes/api.eveonline'
import { get_names_and_categories_by_ids } from '@helpers/api.eveonline/universe'
let names_and_categories:NamesAndCategoriesEvE[]

interface Props {
    character_id?:          number;
    corporation_history?:   CorporationHistoryEvE[];
}

const {
    character_id,
    corporation_history,
} = Astro.props

let corporation_history_eve:CorporationHistoryEvE[]
let get_eve_corporation_history_error
let history_items:CorporationHistoryItem[] | null = null
const names_by_id = {}

if (!corporation_history && character_id) {
    try {
        corporation_history_eve = await get_corporation_history(character_id)

        const history:CorporationHistoryEvE[] = corporation_history ?? (corporation_history_eve ?? [])

        const unique_corporation_ids:number[] = []
        history.forEach( (i) => {
            if (!unique_corporation_ids.includes(i.corporation_id))
                unique_corporation_ids.push(i.corporation_id)
        })

        history_items = history.map( (i, index) => {
            const prev = (index > 0 ? history[index - 1] : null)
            
            const membership_time_text = (prev ?
                from_to_text(lang, i.start_date, prev.start_date)
                :
                from_to_now_text(lang, i.start_date)
            )

            return {
                corporation_id: i.corporation_id,
                membership_time_text: membership_time_text
            }
        })

        names_and_categories = await get_names_and_categories_by_ids(unique_corporation_ids);

        names_and_categories.forEach( (i) => names_by_id[i.id] = i.name )
    } catch (error) {
        get_eve_corporation_history_error = (prod_error_messages() ? t('get_eve_corporation_history_error') : error.message)
    }
}

const CORPORATION_HISTORY_PARTIAL_URL = translatePath('/partials/corporation_history_component/')

const delay = parseInt(Astro.url.searchParams.get('delay') ?? '0')

import Flexblock from '@components/compositions/Flexblock.astro';
import DebugTag from '@components/blocks/DebugTag.astro';
import CorporationBadge from '@components/blocks/CorporationBadge.astro';
import ErrorRefetch from '@components/blocks/ErrorRefetch.astro';
import SectionH4 from '@components/blocks/SectionH4.astro';
---
<SectionH4
    heading_text={t('corporation_history')}
    id="corporation-history"
>
    {!corporation_history && !character_id && 
        <DebugTag set:html={marked.parse(t('corporation_history_params_error'))} />
    }
    {get_eve_corporation_history_error ?
        <ErrorRefetch
            args={{
                partial: `${CORPORATION_HISTORY_PARTIAL_URL}?character_id=${character_id}`,
                message: get_eve_corporation_history_error,
                delay: delay,
                target: '#corporation-history',
            }}
        />
        :
        <Flexblock gap='var(--space-3xs)' class="[ w-full overflow-y-auto max-h-[380px] ]">
            {history_items?.map( (i) => 
                <CorporationBadge
                    corporation={{
                        id: i.corporation_id,
                        name: names_by_id[i.corporation_id] ?? t('unknown_corporation'),
                        size: 'lg',
                        description: i.membership_time_text
                    }}
                />
            )}
        </Flexblock>
    }
</SectionH4>