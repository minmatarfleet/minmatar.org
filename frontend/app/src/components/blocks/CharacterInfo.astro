---
import { getLangFromUrl, useTranslations, useTranslatedPath } from '@i18n/utils';

const lang = getLangFromUrl(Astro.url);
const t = useTranslations(lang);
const translatePath = useTranslatedPath(lang);

import type { User } from '@dtypes/jwt'
import * as jose from 'jose'

const auth_token = Astro.cookies.has('auth_token') ? Astro.cookies.get('auth_token').value : false
const user:User | false = auth_token ? jose.decodeJwt(auth_token) as User : false

interface Props {
    character_id:   number;
}

const {
    character_id,
} = Astro.props


import { get_player_icon } from '@helpers/eve_image_server';
import { prod_error_messages } from '@helpers/env'
import type { CharacterEvE } from '@dtypes/api.eveonline'
import { format_date_time, humanize_date_diff } from '@helpers/date'
import { get_character_by_id } from '@helpers/api.eveonline/characters'
import { get_character_faction, get_race_cover_image, parse_eve_html } from '@helpers/eve'
import { is_html } from '@helpers/string'
import type { CharacterRaces } from '@dtypes/layout_components'

let character:CharacterEvE
let get_character_error:string | false = false
let race:CharacterRaces
let cover:string

import { get_user_permissions } from '@helpers/permissions'
const user_permissions = (user ? await get_user_permissions(user?.username) : [])

try {
    character = await get_character_by_id(character_id);
    
    race = get_character_faction(character.race_id)
    cover = get_race_cover_image(race)
} catch (error) {
    get_character_error = (prod_error_messages() ? t('get_eve_character_error') : error.message)
}

const CHARACTER_MODAL_PARTIAL_URL = translatePath('/partials/character_modal_component')
const delay = parseInt(Astro.url.searchParams.get('delay') ?? '0')

import Wrapper from '@components/compositions/Wrapper.astro'
import Flexblock from '@components/compositions/Flexblock.astro'
import FixedFluid from '@components/compositions/FixedFluid.astro'

import CorporationHistory from '@components/blocks/CorporationHistory.astro'
import CorporationBadge from '@components/blocks/CorporationBadge.astro'
import AllianceBadge from '@components/blocks/AllianceBadge.astro'
import VerticalCenter from '@components/blocks/VerticalCenter.astro'
import ComponentBlock from '@components/blocks/ComponentBlock.astro'
import SectionH4 from '@components/blocks/SectionH4.astro';
import ErrorRefetch from '@components/blocks/ErrorRefetch.astro';
import CharacterLinks from '@components/blocks/CharacterLinks.astro';

import AmarrLogo from '@components/logos/AmarrLogo.astro';
import MinmatarLogo from '@components/logos/MinmatarLogo.astro';
import CaldariLogo from '@components/logos/CaldariLogo.astro';
import GallenteLogo from '@components/logos/GallenteLogo.astro'
---

<Wrapper
    id="character-info"
    class="character-info overflow-y-auto"
    padding_inline='var(--space-s)'
    padding_block='var(--space-s)'
    x-data={`{
        tippy_options: {
            delay: [ 100, 200 ],
            arrow: true,
        },
        init_js() {
            tippy('[data-tippy-content]', this.tippy_options)
        }
    }`}
    x-init="init_js()"
>
    {get_character_error ?
        <ErrorRefetch
            args={{
                partial: `${CHARACTER_MODAL_PARTIAL_URL}?character_id=${character_id}`,
                message: get_character_error,
                delay: delay,
                target: '#character-info',
            }}
        />
        :
        <div class:list={[ 'character-info-cover', cover]} />
        
        <Flexblock gap='var(--space-s)'>
            <FixedFluid
                width='192px'
                class="[ items-center ]"
            >
                <picture>
                    <img loading="lazy" width="192" height="192" src={get_player_icon(character_id, 256)} />
                </picture>
                <div class="[ fluid-fixed-l ]" class="[ w-full ]" style='--fluid-fixed--width: 32px'>
                    <VerticalCenter>
                        <Flexblock gap='var(--space-xs)' class="[ w-full ]">
                            <Flexblock gap='var(--space-3xs)' class="[ w-full ]">
                                <span class="[ text-[var(--highlight)] ]">{character.name}</span>
                                <small>Born: <span data-tippy-content={humanize_date_diff(lang, character.birthday, new Date())}>{format_date_time(lang, character.birthday)}</span></small>
                                <small>Security Status: {character.security_status.toFixed(1)}</small>
                            </Flexblock>
                            <Flexblock gap='var(--space-3xs)' class="[ w-full ]">
                                <CorporationBadge corporation={{
                                    id: character.corporation_id
                                }} />
                                {character.alliance_id &&
                                    <AllianceBadge alliance={{
                                        id: character.alliance_id
                                    }} />
                                }
                                <CharacterLinks
                                    character_id={character_id}
                                    user={user}
                                    user_permissions={user_permissions}
                                />
                            </Flexblock>
                        </Flexblock>
                    </VerticalCenter>
                    <VerticalCenter>
                        <picture>
                            {race === 'amarr' &&
                                <AmarrLogo />
                            }
                            {race === 'minmatar' &&
                                <MinmatarLogo />
                            }
                            {race === 'caldari' &&
                                <CaldariLogo />
                            }
                            {race === 'gallente' &&
                                <GallenteLogo />
                            }
                        </picture>
                    </VerticalCenter>
                </div>
            </FixedFluid>
            
            {character.description !== '' &&
                <ComponentBlock>
                    <SectionH4 heading_text={t('description')}>
                        <div class:list={{ 'eve-html': is_html(character.description) }} set:html={parse_eve_html(character.description)} />
                    </SectionH4>
                </ComponentBlock>
            }
            <ComponentBlock>
                <CorporationHistory character_id={character_id} />
            </ComponentBlock>
        </Flexblock>
    }
</Wrapper>

<style lang="scss">
    .character-info {
        width: 100%;
        max-width: 60rem;
        max-height: min(750px, 88vh);
        height: 100%;
        position: relative;
        background-position: top left;
        background-repeat: no-repeat;
        background-color: var(--background);

        > * {
            position: relative;
            z-index: 1;
        }

        .character-info-cover {
            position: absolute;
            inset: 0;
            z-index: 0;

            &.amarr-cover {
                background-image: url('/images/character-info-amarr-cover.jpg');
            }
        
            &.minmatar-cover {
                background-image: url('/images/character-info-minmatar-cover.jpg');
            }
        
            &.caldari-cover {
                background-image: url('/images/character-info-caldari-cover.jpg');
            }

            &.gallente-cover {
                background-image: url('/images/character-info-gallente-cover.jpg');
            }

            &:before {
                content: ' ';
                inset: 0;
                position: absolute;
                background-image: linear-gradient(
                    to bottom,
                    hsla(0, 0%, 7%, 0) 0%,
                    hsla(0, 0%, 7%, 0.187) 9.8%,
                    hsla(0, 0%, 7%, 0.349) 19.6%,
                    hsla(0, 0%, 7%, 0.488) 29.2%,
                    hsla(0, 0%, 7%, 0.606) 38.6%,
                    hsla(0, 0%, 7%, 0.704) 47.7%,
                    hsla(0, 0%, 7%, 0.784) 56.4%,
                    hsla(0, 0%, 7%, 0.848) 64.5%,
                    hsla(0, 0%, 7%, 0.898) 72.1%,
                    hsla(0, 0%, 7%, 0.936) 78.9%,
                    hsla(0, 0%, 7%, 0.963) 85%,
                    hsla(0, 0%, 7%, 0.981) 90.1%,
                    hsla(0, 0%, 7%, 0.992) 94.3%,
                    hsla(0, 0%, 7%, 0.998) 97.4%,
                    hsl(0, 0%, 7%) 99.3%,
                    hsl(0, 0%, 7%) 100%
                );
            }
        }

        /*> img {
            position: fixed;
            top: 0;
            right: 0;
            display: block;
            width: 100%;
            height: 100%;
            object-fit: cover;
            z-index: -1;
        }*/
    }
</style>