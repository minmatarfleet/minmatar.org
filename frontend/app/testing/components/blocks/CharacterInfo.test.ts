import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { expect, test, vi, beforeEach, afterEach, describe, it } from "vitest";
import CharacterInfo from "@components/blocks/CharacterInfo.astro";

import { get_user_permissions } from "@helpers/permissions";
import { get_character_by_id } from "@helpers/api.eveonline/characters";
import { get_character_faction, get_race_cover_image } from "@helpers/eve";

vi.mock("@helpers/permissions");
vi.mock("@helpers/api.eveonline/characters");
vi.mock("@helpers/eve");

import CorporationHistory from "@components/blocks/CorporationHistory.astro";
import CorporationBadge from "@components/blocks/CorporationBadge.astro";
import AllianceBadge from "@components/blocks/AllianceBadge.astro";
import {
  format_date_time,
  humanize_date_diff,
  is_birthday,
} from "@helpers/date";

vi.mock("@components/blocks/CorporationHistory.astro");
vi.mock("@components/blocks/CorporationBadge.astro");
vi.mock("@components/blocks/AllianceBadge.astro");
vi.mock("@helpers/date");

describe("CharacterInfo", async () => {
  beforeEach(() => {
    // tell vitest we use mocked time
    vi.useFakeTimers();
  });

  afterEach(() => {
    // restoring date after each test run
    vi.useRealTimers();
  });

  it("should dispaly fireworks for birthdays", async () => {
    vi.mocked(get_user_permissions).mockImplementation(async (id) => {
      expect(id).toBe(908);

      return ["cluck", "bagok"];
    });

    vi.mocked(get_character_by_id).mockImplementation(async (id) => {
      expect(id).toBe(908);

      return {
        alliance_id: 12,
        birthday: new Date(1999, 1, 1, 13),
        bloodline_id: 12,
        corporation_id: 1337,
        description: "a goldenretriever like chicken",
        gender: "m",
        name: "Cluckin Kento",
        race_id: 2,
        security_status: -10,
      };
    });

    vi.mocked(get_character_faction).mockImplementation((race) => {
      expect(race).toBe(2);

      return "minmatar";
    });

    vi.mocked(get_race_cover_image).mockImplementation((race) => {
      expect(race).toBe("minmatar");

      return "chicken.gif";
    });

    vi.mocked(humanize_date_diff).mockImplementation((locale, from, to) => {
      return "mock human date";
    });

    vi.mocked(format_date_time).mockImplementation((locale, date) => {
      return "mock date format";
    });

    const date = new Date(2000, 1, 1, 13);
    vi.setSystemTime(date);

    vi.mocked(CorporationHistory).mockResolvedValue("CorporationHistory_mock");
    vi.mocked(CorporationBadge).mockResolvedValue("CorporationBadge_mock");
    vi.mocked(AllianceBadge).mockResolvedValue("AllianceBadge_mock");

    const container = await AstroContainer.create();
    const result = await container.renderToString(CharacterInfo, {
      props: {
        character_id: 908,
      },
    });

    expect(result).toMatchSnapshot();
  });
});
