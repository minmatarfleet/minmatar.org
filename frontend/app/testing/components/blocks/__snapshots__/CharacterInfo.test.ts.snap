// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterInfo > should dispaly fireworks for birthdays 1`] = `
"<div class="wrapper character-info overflow-y-auto" id="character-info" x-data="{
        tippy_options: {
            delay: [ 100, 200 ],
            arrow: true,
        },
        init_js() {
            tippy('[data-tippy-content]', this.tippy_options)
        }
    }" x-init="init_js()" data-astro-cid-z2yiyvvw="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: var(--space-s);--wrapper_padding_inline_end: var(--space-s);--wrapper_padding_block: var(--space-s);--wrapper_max_width: 80rem;"> <div class="character-info-cover chicken.gif" data-astro-cid-z2yiyvvw></div>
        
        <div class="flex-block gap---space-s" data-astro-cid-z2yiyvvw="true" data-mod style="--flex_block_gap: var(--space-s);">  <div class="fixed-fluid [ items-center ]" data-astro-cid-z2yiyvvw="true" data-mod data-astro-cid-34go4nvk style="--fixed_fluid_gap: var(--space-s);--fixed_fluid_width: 192px;--fixed_fluid_breakpoint: 50%;">  <picture data-astro-cid-z2yiyvvw> <img loading="lazy" width="192" height="192" src="https://images.evetech.net/characters/908/portrait?size=256" data-astro-cid-z2yiyvvw> </picture> <div class="[ fluid-fixed-l ]" class="[ w-full ]" style="--fluid-fixed--width: 32px" data-astro-cid-z2yiyvvw> <div class="!flex !items-center" data-astro-cid-z2yiyvvw="true">  <div class="flex-block [ w-full ] gap---space-xs" data-astro-cid-z2yiyvvw="true" data-mod style="--flex_block_gap: var(--space-xs);">  <div class="flex-block [ w-full relative ] gap---space-3xs" data-astro-cid-z2yiyvvw="true" data-mod style="--flex_block_gap: var(--space-3xs);">  <span class="[ text-[var(--highlight)] ]" data-astro-cid-z2yiyvvw>Cluckin Kento</span> <small data-astro-cid-z2yiyvvw>
Born:
<span data-tippy-content="mock human date" data-astro-cid-z2yiyvvw>mock date format</span>  </small> <small data-astro-cid-z2yiyvvw>Security Status: -10.0</small>  </div>  <div class="flex-block [ w-full ] gap---space-3xs" data-astro-cid-z2yiyvvw="true" data-mod style="--flex_block_gap: var(--space-3xs);">  CorporationBadge_mock AllianceBadge_mock<div class="flex-inline" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-3xs);--flex_inline_justification: flex-start;"> <a data-tippy-content="Zkillboard" class="[ w-[32px] neocom-button neocom-button-focus ]" target="_blank" rel="nofollow noopener noreferrer" x-data="{
            show_external_link_disclaimer() {
                show_alert_dialog({
                    title: 'Youâ€™re Leaving Our Site',
                    content: 'The link you clicked will take you to an external website.',
                }).then( (accepted) => accepted &#38;&#38; window.location.href = &#34;undefined&#34; )
            }
        }" x-on:click.prevent="show_external_link_disclaimer"> <img loading="lazy" class="[ max-w-[100%] h-auto ]" src="/images/icons/48px-Combatlog.png" width="48" height="48" alt="EVE Online Combat Log Icon"> </a>  </div>    </div>   </div>   </div> <div class="!flex !items-center" data-astro-cid-z2yiyvvw="true">  <picture data-astro-cid-z2yiyvvw>  <svg class="[ w-full ]" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="-0.393 -0.315 143 143" xml:space="preserve"> <defs></defs> <g> <path d="M112.727,14.19L82.745,69.283c-5.087-4.35-11.397-6.926-18.234-6.926c-16.848,0-30.507,15.643-30.507,34.94    c0,13.53,7.927,24.501,17.206,30.84c8.531,5.829,18.532,8.477,29.511,8.477c33.695,0,61.012-28.25,61.012-63.098    C141.732,49.158,130.313,27.568,112.727,14.19 M65.053,54.267c5.418,0,10.604,1.058,15.388,2.991l25.272-47.77    C95.424,3.448,83.533,0,70.865,0c-8.779,0-17.184,1.661-24.943,4.689l16.473,49.673C63.275,54.305,64.16,54.267,65.053,54.267     M53.351,55.982L37.629,8.576C23.297,16.487,11.891,29.351,5.541,44.971l29.92,21.715C40.469,61.717,46.576,57.998,53.351,55.982     M29.668,73.721L2.524,54.02C0.881,60.229,0,66.765,0,73.516c0,7.693,1.145,15.108,3.258,22.078h18.574    C22.662,87.491,25.446,80.021,29.668,73.721 M21.732,104.688H6.68c7.765,17.178,21.738,30.734,38.932,37.539    C32.411,135.173,23.048,121.157,21.732,104.688" fill="#653834"></path> </g> </svg>   </picture>  </div> </div>  </div>   <div class="wrapper [ card component ] light-transparency" data-astro-cid-z2yiyvvw="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: var(--space-l-xl);--wrapper_padding_inline_end: var(--space-l-xl);--wrapper_padding_block: var(--component-padding-block);--wrapper_max_width: 100%;">   <div class="flex-block gap---space-xs" data-astro-cid-z2yiyvvw="true" data-mod style="--flex_block_gap: var(--space-xs);">  <h4>Description</h4>  <div data-astro-cid-z2yiyvvw></div>   </div>    </div>  <div class="wrapper [ card component ] light-transparency" data-astro-cid-z2yiyvvw="true" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: var(--space-l-xl);--wrapper_padding_inline_end: var(--space-l-xl);--wrapper_padding_block: var(--component-padding-block);--wrapper_max_width: 100%;">   CorporationHistory_mock   </div>    </div>  </div>   "
`;
